pkg/apis/node/types.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// RuntimeClass defines a class of container runtime supported in the cluster.
+// The RuntimeClass is used to select or configure the container runtime that is
+// used to run all containers in a pod. RuntimeClasses are (currently) manually
+// defined by a user or cluster provisioner, and referenced in the PodSpec. The
+// Kubelet is responsible for resolving the RuntimeClassName reference before
+// running the pod.  For more details, see
+// https://git.k8s.io/enhancements/keps/sig-node/runtime-class.md
+type RuntimeClass struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ObjectMeta
+
+	// Specification of the RuntimeClass
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	Spec RuntimeClassSpec
+}
+
+// RuntimeClassSpec is a specification of a RuntimeClass. It contains parameters
+// that required to describe the RuntimeClass to the Container Runtime Interface
+// (CRI) implementation, as well as any other components that need to understand
+// how the pod will be run. The RuntimeClassSpec should be treated as immutable.
+type RuntimeClassSpec struct {
+	// RuntimeHandler specifies the underlying runtime and configuration that the
+	// CRI implementation will use to handle pods of this class. The possible
+	// values are specific to the node configuration & CRI implementation.  It is
+	// assumed that all handlers are available on every node, and handlers of the
+	// same name are equivalent on every node.
+	// For example, a handler called "runc" might specify that the runc OCI
```

Comment:
>  ... and configuration that the CRI implementation **will use** ...

The point is that the CRI implementation is the component that actually acts on this value, not the kubelet.

Do you have any suggestions for how to describe this more clearly? Should I just drop the reference to the CRI implementation? I wanted to clarify that this isn't selecting between CRI-o and containerd though, but rather between runc and runsc, etc.

Reply:
What stops someone from using this field for a different purpose that what's intended coz it's upto the CRI runtime to interpret this field? Part of the confusion is due to the use of the word "runtime" we discussed in a previous comment.
Maybe you can drop "configuration" from `RuntimeHandler specifies the underlying runtime and configuration that the CRI implementation will use` to ensure this field is clearly expected to be used to specify the underlying `sandboxing` solution?
Re-reading the current version, I feel less confused. Maybe we can wait for @thockin to do a pass over this PR and see if he finds the comments unambiguous.

Reply:
> What stops someone from using this field for a different purpose that what's intended coz it's upto the CRI runtime to interpret this field? 

Nothing, it could be completely abused. Maybe we'll need a concept of runtime conformance eventually?

> Maybe you can drop "configuration"

I don't want to exclude running the same "runtime binary" (e.g. runsc) with different configurations. For example, if we wanted to have a "gvisor" and "gvisor-lite" runtimeclass, where the "lite" version sacrificed some extra defenses in the name of performance (e.g. use the host TCP stack).

Reply:
> Nothing, it could be completely abused. Maybe we'll need a concept of runtime conformance eventually?

Hmm.. Probably not coz conformance would only make sense at the "Runtime Class" naming level right?

Reply:
Also, do you expect users to want to pass more arbitrary configuration information through Runtime Class object? The reason I ask is if the Runtime Handler fields should end up becoming an arbitrary list which is only interpretable by runtimes.

Reply:
nvm, I see below that you intend this not be used to pass arbitrary information by controlling the naming spec